src/cli/crd/generate.test.ts

```diff
@@ -0,0 +1,202 @@
+    });
+  });
+
+  describe("getAPIVersions", () => {
+    it("should return directories from the api root", () => {
+      (fs.readdirSync as jest.Mock).mockReturnValue(["v1", "v2"]);
+      (fs.statSync as jest.Mock).mockImplementation(p => ({
+        isDirectory: () => typeof p === "string" && (p.endsWith("v1") || p.endsWith("v2")),
+      }));
+
+      const versions = getAPIVersions("/api");
+      expect(versions).toEqual(["v1", "v2"]);
+    });
+
+    it("should ignore non-directory entries", () => {
+      (fs.readdirSync as jest.Mock).mockReturnValue(["v1", "README.md"]);
+      (fs.statSync as jest.Mock).mockImplementation(p => ({
+        isDirectory: () => typeof p === "string" && p.endsWith("v1"),
+      }));
+
+      const versions = getAPIVersions("/api");
+      expect(versions).toEqual(["v1"]);
+    });
+  });
+
+  describe("loadVersionFiles", () => {
+    it("should load only TypeScript files from version directory", () => {
+      const project = new Project();
+      const addSpy = jest.spyOn(project, "addSourceFilesAtPaths").mockReturnValue([]); // don't resolve real files
+
+      (fs.readdirSync as jest.Mock).mockReturnValue(["foo.ts", "bar.js", "baz.ts"]);
+
+      const result = loadVersionFiles(project, "/api/v1");
+
+      expect(addSpy).toHaveBeenCalledWith(["/api/v1/foo.ts", "/api/v1/baz.ts"]);
+      expect(result).toEqual([]); // we mocked return value
+    });
+  });
+
+  describe("extractDetails", () => {
+    const createProjectWithFile = (name: string, content: string) => {
+      const project = new Project();
+      return project.createSourceFile(name, content);
+    };
+
+    it("should extract plural, scope, and shortName from the details object", () => {
+      const file = createProjectWithFile(
+        "temp.ts",
+        `
+        const details = { plural: "widgets", scope: "Namespaced", shortName: "wd" };
+      `,
+      );
+
+      const details = extractDetails(file);
+      expect(details).toEqual({
+        plural: "widgets",
+        scope: "Namespaced",
+        shortName: "wd",
+      });
+    });
+
+    it("should throw if 'scope' is invalid", () => {
+      const file = createProjectWithFile(
+        "bad.ts",
+        `
+        const details = { plural: "widgets", scope: "BadScope", shortName: "wd" };
+      `,
+      );
+
+      expect(() => extractDetails(file)).toThrow(
+        `'scope' must be either "Cluster" or "Namespaced", got "BadScope"`,
+      );
+    });
+
+    it("should throw if details variable is missing", () => {
+      const file = createProjectWithFile("noDetails.ts", `const somethingElse = {};`);
+      expect(() => extractDetails(file)).toThrow(`Missing 'details' variable declaration.`);
+    });
+  });
+
+  describe("processSourceFile", () => {
+    const createProjectWithFile = (name: string, content: string) => {
+      const project = new Project();
+      return project.createSourceFile(name, content);
+    };
```

Comment:
This looks identical to the one in `extractDetails`, thoughts on extracting this to a common generator function?

Reply:
see #2066